File: DistributeWater.java
//TC : O((V + E) *  log(V + E) ) // V - total no.of houses and E - total edges/pipes between all houses
//Sc : O(V + E)  

class Solution {
    int[] unionFind;
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        if(n == 0 || wells == null)   return 0;
        
        unionFind = new int[n+1];
        
        
        for(int i = 0; i <= n; i++){
            unionFind[i] = i;
        }
        
        List<int[]> costs = new ArrayList<>();
        
        for(int w = 0; w < wells.length; w++){
            costs.add(new int[]{0, w+1, wells[w] });
        }
        for(int[] p : pipes){
            costs.add(p);
        }
        
        Collections.sort(costs, (a,b) -> a[2] - b[2]);
        
        int cost = 0;
        for(int[] curr : costs){
            int px = find(curr[0]);
            int py = find(curr[1]);
            if(px != py){
                cost += curr[2];
                unionFind[px] = py; // or unionFind[py] = px;
            }
        }
        return cost;
    }
    
    public int find(int x){//find parent of node
        if(x != unionFind[x]){
            unionFind[x] = find(unionFind[x]);
        }
        return unionFind[x];
    }
}

File: distributeWater.cpp
//Time: O(V + ElogE) Because of sorting 
//Space: 0(V+E) where V and E are the vertices and edges 
using Edge = vector<int>; //{src,dest,cost}
bool comparator(Edge a ,Edge b){
    return a[2] <= b[2];
}
class Solution {
    vector<Edge> edges;
    vector<int> unionFind;
public:
    int find(int val){
        if(unionFind[val] == val) return val;
        return find(unionFind[val]);
        
    }
    int minCostToSupplyWater(int n, vector<int>& wells, vector<vector<int>>& pipes) {
        unionFind.resize(n+1);
        int minCount= 0;
        for(int i = 0; i <= n; i++) unionFind[i] = i;
        for(Edge pipe : pipes) edges.push_back(pipe);
        for(int i = 1; i <= wells.size(); i++){
            edges.push_back({0,i,wells[i-1]});
        }
        sort(edges.begin(),edges.end(),comparator);
        for(Edge ed : edges){
            int px = find(ed[0]);
            int py = find(ed[1]);
            if(px != py){
                minCount += ed[2];
                unionFind[px] = py;
            }
        }
        return minCount;
    }
};